engine/src/main/java/org/pentaho/di/trans/SingleThreadedTransExecutor.java

```diff
@@ -0,0 +1,432 @@
+    // using a threshold.
+    //
+    // After this many iterations enable trimming inner iteration
+    // window on no change being detected.
+    //
+    int windowShrinkThreshold = (int) Math.round( stepsSize * 0.75 );
+
+    // give ourselves some room to sort big lists. the window threshold should
+    // stop us before reaching this anyway.
+    //
+    int totalIterations = stepsSize * 2;
+    int actualIterations = 0;
+
+    boolean isBefore = false;
+    boolean forwardChange = false;
+    boolean backwardChange = false;
+
+    boolean lastForwardChange = true;
+    boolean keepSortingForward = true;
+
+    StepMetaDataCombi one = null;
+    StepMetaDataCombi two = null;
+
+    StringBuilder tLogString = new StringBuilder(); // this helps group our
+                                                    // output so other threads
+                                                    // don't get logs in our
+                                                    // output.
+    tLogString.append( "-------------------------------------------------------" ).append( "\n" );
+    tLogString.append( "--SingleThreadedTransExecutor.sortSteps(cocktail)" ).append( "\n" );
+    tLogString.append( "--Trans: " ).append( trans.getName() ).append( "\n" );
+    tLogString.append( "-" ).append( "\n" );
+
+    long startTime = System.currentTimeMillis();
+
+    for ( int x = 0; x < totalIterations; x++ ) {
+
+      // Go forward through the list
+      //
+      if ( keepSortingForward ) {
+        for ( int y = stepsMinSize; y < stepsSize - 1; y++ ) {
+          one = steps.get( y );
+          two = steps.get( y + 1 );
+          isBefore = trans.getTransMeta().findPrevious( one.stepMeta, two.stepMeta );
+          if ( isBefore ) {
+            // two was found to be positioned BEFORE one so we need to
+            // switch them...
+            //
+            steps.set( y, two );
+            steps.set( y + 1, one );
+            forwardChange = true;
+
+          }
+        }
+      }
+
+      // Go backward through the list
+      //
+      for ( int z = stepsSize - 1; z > stepsMinSize; z-- ) {
+        one = steps.get( z );
+        two = steps.get( z - 1 );
+
+        isBefore = trans.getTransMeta().findPrevious( one.stepMeta, two.stepMeta );
+        if ( !isBefore ) {
+          // two was found NOT to be positioned BEFORE one so we need to
+          // switch them...
+          //
+          steps.set( z, two );
+          steps.set( z - 1, one );
+          backwardChange = true;
+        }
+      }
+
+      // Shrink stepsSize(max) if there was no forward change
+      //
+      if ( x > windowShrinkThreshold && !forwardChange ) {
+
+        // should we keep going? check the window size
+        //
+        stepsSize--;
+        if ( stepsSize <= stepsMinSize ) {
+          if ( testing ) {
+            tLogString.append( String.format( "stepsMinSize:%s  stepsSize:%s", stepsMinSize, stepsSize ) );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code uses &#39;%s&#39; to format int: stepsSize (declared at line 193), stepsMinSize (declared at line 105). This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 196. Consider formatting this data with &#39;%d&#39; instead.